[SPARK-8119] HeartbeatReceiver should replace executors, not kill

andrewor14 opened the pull request:
**Symptom.** If an executor in an application times out, `HeartbeatReceiver` attempts to kill it. After this happens, however, the application never gets an executor back even when there are cluster resources available.

**Cause.** The issue is that `sc.killExecutor` automatically assumes that the application wishes to adjust its resource requirements permanently downwards. This is not the intention in `HeartbeatReceiver`, however, which simply wants a replacement for the expired executor.

**Fix.** Differentiate between the intention to kill and the intention to replace an executor with a fresh one. More details can be found in the commit message.

Commit:
Differentiate between kill and replace

The issue is that sc.killExecutors automatically assumes that
the application wishes to adjust its resource requirements
permanently downwards. This is not the intention in
HeartbeatReceiver, however, which simply wants a replacement
for the expired executor.

andrewor14 commented:
This is an alternative to #6662. @vanzin @SaintBacchus

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36067 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36067/consoleFull) for   PR 7107 at commit [`18b110e`](https://github.com/apache/spark/commit/18b110e6dee9d8cfe7f2cf5c0b3133bce2bbf551).

SparkQA commented:
[Test build #36067 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36067/console) for   PR 7107 at commit [`18b110e`](https://github.com/apache/spark/commit/18b110e6dee9d8cfe7f2cf5c0b3133bce2bbf551).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

andrewor14 commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36070 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36070/consoleFull) for   PR 7107 at commit [`18b110e`](https://github.com/apache/spark/commit/18b110e6dee9d8cfe7f2cf5c0b3133bce2bbf551).

SparkQA commented:
[Test build #36070 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36070/console) for   PR 7107 at commit [`18b110e`](https://github.com/apache/spark/commit/18b110e6dee9d8cfe7f2cf5c0b3133bce2bbf551).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

andrewor14 commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36183 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36183/consoleFull) for   PR 7107 at commit [`18b110e`](https://github.com/apache/spark/commit/18b110e6dee9d8cfe7f2cf5c0b3133bce2bbf551).

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -394,6 +394,27 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
+   *
+   * Return whether the request is received.
+   */
+  def killAndReplaceExecutor(executorId: String): Boolean = synchronized {
```
Perhaps instead of adding this method, add a boolean flag to `killExecutors` above?

(Or, because of `ExecutorAllocationClient`, create a private method that takes the list of executors and a boolean saying whether to update the target executor count, and call it from both of these methods.)

andrewor14 replied:
I was planning to do that but it technically changes public API. The methods in `ExecutorAllocationClient` are actually public because `SparkContext` extends it, so we actually can't change the signature there.

vanzin replied:
Yes, but you can follow the second suggestion (private method called by the other two).

andrewor14 replied:
OK, sounds good. I didn't do it initially because there isn't really a great name for the private method (we already have `doKillExecutors`). I'll figure something out.

vanzin replied:
In case you haven't done it yet, you might be able to just call this method from `killExecutors`.

vanzin commented:
Approach looks sane, it just feels like both implementations could share more code.

SparkQA commented:
**[Test build #36183 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36183/console)**     for PR 7107 at commit [`18b110e`](https://github.com/apache/spark/commit/18b110e6dee9d8cfe7f2cf5c0b3133bce2bbf551)     after a configured wait of `175m`.

AmplabJenkins commented:
Merged build finished. Test PASSed.

andrewor14 commented:
This is awaiting the clean up in `HeartbeatReceiverSuite` in #7173.

AmplabJenkins commented:
Merged build triggered.

andrewor14 commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37276 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37276/consoleFull) for   PR 7107 at commit [`18b110e`](https://github.com/apache/spark/commit/18b110e6dee9d8cfe7f2cf5c0b3133bce2bbf551).

SparkQA commented:
[Test build #37276 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37276/console) for   PR 7107 at commit [`18b110e`](https://github.com/apache/spark/commit/18b110e6dee9d8cfe7f2cf5c0b3133bce2bbf551).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

xuchenCN commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -394,6 +394,27 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
+  def killAndReplaceExecutor(executorId: String): Boolean = synchronized {
+    logInfo(s"Requesting to replace executor $executorId")
+    if (executorDataMap.contains(executorId)) {
+      executorsPendingToRemove += executorId
```
@andrewor14 If CoarseGrainedSchedulerBackend.removeExecutor was happened before HeartbeatReceiver.killExecutorThread

andrewor14 replied:
If we already called `removeExecutor` then `executorDataMap` will not contain `executorId`, so we won't even get this `if` case. I believe the scenario you described cannot happen.

Commit:
Reuse more code in scheduler backend

Commit:
Add regression test for SPARK-8119

This issue is actually somewhat difficult to test. There are many
components in the picture and each one has to be mocked out. Many
of the executor allocation requests are also done asynchronously
so we need to find ways to block on these requests.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37542 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37542/consoleFull) for   PR 7107 at commit [`1cd2cd7`](https://github.com/apache/spark/commit/1cd2cd79fefa010de15203ed46ccca5a9667cf01).

andrewor14 commented:
@vanzin I updated this with your suggestion.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #1088 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1088/consoleFull) for   PR 7107 at commit [`1cd2cd7`](https://github.com/apache/spark/commit/1cd2cd79fefa010de15203ed46ccca5a9667cf01).

SparkQA commented:
[Test build #1090 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1090/consoleFull) for   PR 7107 at commit [`1cd2cd7`](https://github.com/apache/spark/commit/1cd2cd79fefa010de15203ed46ccca5a9667cf01).

SparkQA commented:
[Test build #1092 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1092/consoleFull) for   PR 7107 at commit [`1cd2cd7`](https://github.com/apache/spark/commit/1cd2cd79fefa010de15203ed46ccca5a9667cf01).

vanzin commented:
LGTM.

SparkQA commented:
[Test build #37542 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37542/console) for   PR 7107 at commit [`1cd2cd7`](https://github.com/apache/spark/commit/1cd2cd79fefa010de15203ed46ccca5a9667cf01).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class UnresolvedAttribute(nameParts: Seq[String]) extends Attribute `
  * `abstract class Star extends LeafExpression with NamedExpression `
  * `case class UnresolvedAlias(child: Expression) extends UnaryExpression with NamedExpression `
  * `case class SortOrder(child: Expression, direction: SortDirection) extends UnaryExpression `
  * `trait AggregateExpression extends Expression `
  * `trait PartialAggregate extends AggregateExpression `
  * `case class Min(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Max(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Count(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Average(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Sum(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class SumDistinct(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class First(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Last(child: Expression) extends UnaryExpression with PartialAggregate `
  * `trait Generator extends Expression `
  * `case class Explode(child: Expression) extends UnaryExpression with Generator `
  * `trait NamedExpression extends Expression `
  * `abstract class Attribute extends LeafExpression with NamedExpression `
  * `case class PrettyAttribute(name: String) extends Attribute `
  * `abstract class LeafNode extends LogicalPlan `
  * `abstract class UnaryNode extends LogicalPlan `

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
[Test build #1088 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1088/console) for   PR 7107 at commit [`1cd2cd7`](https://github.com/apache/spark/commit/1cd2cd79fefa010de15203ed46ccca5a9667cf01).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class UnresolvedAttribute(nameParts: Seq[String]) extends Attribute `
  * `abstract class Star extends LeafExpression with NamedExpression `
  * `case class UnresolvedAlias(child: Expression) extends UnaryExpression with NamedExpression `
  * `case class SortOrder(child: Expression, direction: SortDirection) extends UnaryExpression `
  * `trait AggregateExpression extends Expression `
  * `trait PartialAggregate extends AggregateExpression `
  * `case class Min(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Max(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Count(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Average(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Sum(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class SumDistinct(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class First(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Last(child: Expression) extends UnaryExpression with PartialAggregate `
  * `trait Generator extends Expression `
  * `case class Explode(child: Expression) extends UnaryExpression with Generator `
  * `trait NamedExpression extends Expression `
  * `abstract class Attribute extends LeafExpression with NamedExpression `
  * `case class PrettyAttribute(name: String) extends Attribute `
  * `abstract class LeafNode extends LogicalPlan `
  * `abstract class UnaryNode extends LogicalPlan `

SparkQA commented:
[Test build #1092 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1092/console) for   PR 7107 at commit [`1cd2cd7`](https://github.com/apache/spark/commit/1cd2cd79fefa010de15203ed46ccca5a9667cf01).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `abstract class StandaloneRecoveryModeFactory(conf: SparkConf, serializer: Serializer) `
  * `class RFormula(override val uid: String)`
  * `case class UnresolvedAttribute(nameParts: Seq[String]) extends Attribute `
  * `abstract class Star extends LeafExpression with NamedExpression `
  * `case class UnresolvedAlias(child: Expression) extends UnaryExpression with NamedExpression `
  * `abstract class LeafExpression extends Expression `
  * `abstract class UnaryExpression extends Expression `
  * `abstract class BinaryExpression extends Expression `
  * `case class SortOrder(child: Expression, direction: SortDirection) extends UnaryExpression `
  * `trait AggregateExpression extends Expression `
  * `trait PartialAggregate extends AggregateExpression `
  * `case class Min(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Max(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Count(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Average(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Sum(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class SumDistinct(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class First(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Last(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Pmod(left: Expression, right: Expression) extends BinaryArithmetic `
  * `      final class SpecificRow extends $`
  * `trait Generator extends Expression `
  * `case class Explode(child: Expression) extends UnaryExpression with Generator `
  * `trait NamedExpression extends Expression `
  * `abstract class Attribute extends LeafExpression with NamedExpression `
  * `case class PrettyAttribute(name: String) extends Attribute `
  * `case class Length(child: Expression) extends UnaryExpression with ExpectsInputTypes `
  * `case class FormatNumber(x: Expression, d: Expression)`
  * `abstract class LeafNode extends LogicalPlan `
  * `abstract class UnaryNode extends LogicalPlan `
  * `abstract class BinaryNode extends LogicalPlan `

SparkQA commented:
[Test build #1090 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1090/console) for   PR 7107 at commit [`1cd2cd7`](https://github.com/apache/spark/commit/1cd2cd79fefa010de15203ed46ccca5a9667cf01).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class UnresolvedAttribute(nameParts: Seq[String]) extends Attribute `
  * `abstract class Star extends LeafExpression with NamedExpression `
  * `case class UnresolvedAlias(child: Expression) extends UnaryExpression with NamedExpression `
  * `case class SortOrder(child: Expression, direction: SortDirection) extends UnaryExpression `
  * `trait AggregateExpression extends Expression `
  * `trait PartialAggregate extends AggregateExpression `
  * `case class Min(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Max(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Count(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Average(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Sum(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class SumDistinct(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class First(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Last(child: Expression) extends UnaryExpression with PartialAggregate `
  * `trait Generator extends Expression `
  * `case class Explode(child: Expression) extends UnaryExpression with Generator `
  * `trait NamedExpression extends Expression `
  * `abstract class Attribute extends LeafExpression with NamedExpression `
  * `case class PrettyAttribute(name: String) extends Attribute `
  * `case class Length(child: Expression) extends UnaryExpression with ExpectsInputTypes `
  * `case class FormatNumber(x: Expression, d: Expression)`
  * `abstract class LeafNode extends LogicalPlan `
  * `abstract class UnaryNode extends LogicalPlan `

xuchenCN commented:
+1 LGTM

andrewor14 commented:
Thanks @xuchenCN and @vanzin. I'm merging this into master. A backport for 1.4.2 will come later.
